File: src/features/typing/data/ts/flatten_binary_tree_to_linked_list.ts
class TreeNode {
    val: number;
    left: TreeNode | null;
    right: TreeNode | null;
    constructor(val = 0, left: TreeNode | null = null, right: TreeNode | null = null) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

function flatten(root: TreeNode | null): void {
    let prev: TreeNode | null = null;
    const flattenNode = (node: TreeNode | null): void => {
        if (node === null) {
            return;
        }
        flattenNode(node.right);
        flattenNode(node.left);
        node.right = prev;
        node.left = null;
        prev = node;
    };
    flattenNode(root);
}

File: src/features/typing/data/go/flatten_binary_tree_to_linked_list.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func flatten(root *TreeNode) {
	var prev *TreeNode
	var flattenNode func(node *TreeNode)
	flattenNode = func(node *TreeNode) {
		if node == nil {
			return
		}
		flattenNode(node.Right)
		flattenNode(node.Left)
		node.Right = prev
		node.Left = nil
		prev = node
	}
	flattenNode(root)
}
